### ISSUE TITLE
question about self-attention in cut3r

### ISSUE THREAD

**Issue by ou524u** [Inception3D/TTT3R#3]
Hi, thanks for your great work!

I have a question regarding the illustration in your paper.

In the paper, the figure below illustrates how CUT3R works, and TTT3R is built upon that:
<img width="1161" height="394" alt="Image" src="https://github.com/user-attachments/assets/efd760cd-fe6c-4379-a9b4-06aa0bc859bb" />

But I have a question — as I understand it, in CUT3R they are using one Transformer with self-attention over concat([s_{t-1}, x_t]), so instead of a pure cross-attention formulation as:
![formula](https://latex.codecogs.com/svg.image?\mathbf{S}_t=\mathbf{S}_{t-1}+\mathrm{softmax}\left(\frac{\mathbf{Q}_{\mathbf{S}_{t-1}}\mathbf{K}_{\mathbf{X}_t}^\top}{\sqrt{d_k}}\right)\mathbf{V}_{\mathbf{X}_t})

it should actually be formulated as:
![formula](https://latex.codecogs.com/svg.image?\mathbf{S}_t=\mathbf{S}_{t-1}+\mathrm{softmax}\left(\frac{\mathbf{Q}_{\mathbf{S}_{t-1}}[\mathbf{K}_{\mathbf{S}_{t-1}};\mathbf{K}_{\mathbf{X}_t}]^\top}{\sqrt{d_k}}\right)[\mathbf{V}_{\mathbf{S}_{t-1}};\mathbf{V}_{\mathbf{X}_t}])

Am I wrong?
And how would that affect TTT3R, since your \beta_t seems to be based on the cross-attention formulation?
Looking forward to your explanation on this issue.
Thanks again for your impressive work!

**Comment by rover-xingyu** [comment]
Hi @ou524u, thanks for your interest! 

You are right: following DUSt3R, CUT3R sequentially applies self-attention, followed by cross-attention. 

In TTT3R, we focus on analyzing the cross-attention component (aligned with findings from [Easi3R](https://easi3r.github.io/)), and our β_t is defined using the cross-attention weights. We have not yet found a specific role of the self-attention, a deeper investigation of self-attention effects is left for future work.

**Comment by ou524u** [comment]
Hi, thanks for your quick reply!

Since β_t is defined using the cross-attention weights, I’m wondering — is β_t also applied to the self-attention?
It seems from your code that the self-attention part remains unchanged  ([link](https://github.com/Inception3D/TTT3R/blob/cc6142bd9c15df57817a5c7a4797c163e46fa221/src/dust3r/model.py#L903)), am I understanding this correctly?

If possible, could you share a small code snippet showing how to plug your modification into CUT3R? It looks simple and quite effective, where a minimal example would make it much easier to try out.

Thanks again for your time and excellent work!

**Comment by rover-xingyu** [comment]
Yes, the β_t is applied to the block output after self‑attention and cross‑attention are executed sequentially. 

The main modifications are [attention capture](https://github.com/Inception3D/TTT3R/blob/cc6142bd9c15df57817a5c7a4797c163e46fa221/src/dust3r/blocks.py#L245) and  [test‑time training with learning rate](https://github.com/Inception3D/TTT3R/blob/cc6142bd9c15df57817a5c7a4797c163e46fa221/src/dust3r/model.py#L903).

**Comment by ou524u** [comment]
Thanks for your reply!
